addons/mail/static/src/messaging/entity/entity/entity.js

```diff
@@ -0,0 +1,824 @@
+
+        /**
+         * @static
+         * @param {string|mail.messaging.entity.Entity|undefined} entityOrLocalId
+         * @returns {mail.messaging.entity.Entity|undefined}
+         */
+        static get(entityOrLocalId) {
+            const { state } = this.env.store;
+            if (entityOrLocalId === undefined) {
+                return undefined;
+            }
+            const entity = state.entities[
+                entityOrLocalId.isEntity
+                    ? entityOrLocalId.localId
+                    : entityOrLocalId
+            ];
+            if (!(entity instanceof this) && entity !== this) {
+                return;
+            }
+            return entity;
+        }
+
+        static link(data) {
+            this.__link(this, data);
+        }
+
+        /**
+         * @static
+         * @param {Object} data
+         * @param {any} data.id
+         * @returns {mail.messaging.entity.Entity}
+         */
+        static insert(data) {
+            let entity = this.fromId(data.id);
+            if (!entity) {
+                entity = this.create(data);
+            } else {
+                entity.update(data);
+            }
+            return entity;
+        }
+
+        /**
+         * @throws {Error} when called with non singleton classes
+         * @returns {mail.messaging.entity.Entity|undefined}
+         */
+        static get instance() {
+            if (!this.isSingleton) {
+                throw new Error(`Cannot get instance of non-singleton Entity class "${this.name}"`);
+            }
+            return this.all[0];
+        }
+
+        /**
+         * @static
+         * @returns {boolean}
+         */
+        static get isEntity() {
+            return true;
+        }
+
+        /**
+         * @static
+         * @returns {boolean}
+         */
+        static get isEntityInstance() {
+            return false;
+        }
+
+        /**
+         * @static
+         * @returns {boolean}
+         */
+        static get isEntityClass() {
+            return true;
+        }
+
+        /**
+         * @static
+         * @param {Object|string} data
+         */
+        static unlink(data) {
+            this.__unlink(this, data);
+        }
+
+        /**
+         * @static
+         * @param {Object} data
+         */
+        static update(data) {
+            this.__update(this, data);
+        }
+
+        delete() {
+            this.constructor.__delete(this);
+        }
+
+        /**
+         * @returns {boolean}
+         */
+        get isEntity() {
+            return true;
+        }
+
+        /**
+         * @returns {boolean}
+         */
+        get isEntityClass() {
+            return false;
+        }
+
+        /**
+         * @returns {boolean}
+         */
+        get isEntityInstance() {
+            return true;
+        }
+
+        /**
+         * @param {Object} data
+         */
+        link(data) {
+            this.constructor.__link(this, data);
```

Comment:
you need to explain in more details the static stuff, this seems very unnecessary complex